internal/clients/telegram/client.go

```diff
@@ -0,0 +1,282 @@
+	c.wg.Add(1)
+	defer c.wg.Done()
+	// TODO: Hold back updates that could cause race conditions in parallel processing
+	defer close(stateCh)
+	for upd := range updateCh {
+		stateCh <- updateHandler{
+			processor: upd,
+			state:     c.takeState(),
+		}
+	}
+}
+
+/*
+TODO: should return state for the update that will be processed.
+Should also block if the state is taken already
+(Should take some args to look up the state for the conversation)
+*/
+func (c *Client) takeState() ConversationStateHandler {
+	return &rootConversationState{}
+}
+
+// TODO: Stores the state and allows other threads to take it via takeState
+func (c *Client) releaseState(ConversationStateHandler) {}
+
+/*
+processUpdates method should be run in a goroutine and will process updates
+that come through the channel.
+
+Stop this goroutine by closing the channel
+*/
+func (c *Client) processUpdates(updateQueue <-chan updateHandler) {
+	c.wg.Add(1)
+	defer c.wg.Done()
+
+	for update := range updateQueue {
+		state, actions := update.processor.processTelegramUpdate(update.state)
+		for _, action := range actions {
+			endpoint, query := action.telegramBotAction()
+
+			req := c.mustGetRequest(endpoint, query, nil)
+			_, err := doApiRequest[struct{}](c, req, "/"+endpoint)
+			if err != nil {
+				log.Printf("Error while performing bot action: %s\n", err)
+			}
+		}
+		c.releaseState(state)
+	}
+}
+
+// Makes a GET method request from components. Panics if the request cannot be created
+func (c *Client) mustGetRequest(endpoint string, query url.Values, body io.Reader) *http.Request {
+	u := url.URL{
+		Scheme:   "https",
+		Host:     c.host,
+		Path:     path.Join(c.basePath, endpoint),
+		RawQuery: query.Encode(),
+	}
+	req, err := http.NewRequest(http.MethodGet, u.String(), body)
+	if err != nil {
+		// panic because a request should always be valid
+		log.Fatal("Error: Request should always be valid, %w", err)
+	}
+	return req
+}
+
+/*
+Performs the Telegram API request, does error handling, wraps errors and then returns the payload of the request
+
+`logID` is the string that will be in the logs and error messages and is used to track what method failed. Can be arbitrary.
+
+# Panics
+
+Panics if the status code is 401 Unauthorized which is usually caused by an invalid bot token.
+
+Or if the telegram API sends a "Migrate to chat ID" error, which is unsupported as of now and could cause weird bugs.
+
+Will also panic if the telegram API returns an error many timess in a row (probably 10 times, but look at the source to see
+the exact value.
+*/
+func doApiRequest[T any](c *Client, req *http.Request, logID string) (*T, error) {
+	const retry = 3
+	for i := 0; i < retry; i++ {
+		resp, err := c.client.Do(req)
+		if err != nil {
+			return nil, fmt.Errorf("Network error: (%s): %s", logID, err)
+		}
+
+		body, err := io.ReadAll(resp.Body)
+		resp.Body.Close()
+		if err != nil {
+			return nil, fmt.Errorf("Could not read responce body: (%s): %s", logID, err)
```

Comment:
please use lowercase for error:
fmt.Errorf("could not read responce body: (%s): %s", logID, err)